File: cogops/verifier/policy.py
"""
cogops/verifier/policy.py

Decision policy that turns NLI verdicts into a final answer + event stream.

Policies:
  - "redact":  not_entailed sentences are replaced with [তথ্য যাচাইযোগ্য নয়];
               partial → keep + emit warning; entailed → keep.
               If >50% of cited sentences are not_entailed, escalate to refuse.
  - "refuse":  any single not_entailed → replace the whole answer with the
               static refusal template.
  - "warn":    keep the answer as-is, emit warnings for non-entailed. Used
               automatically when the verifier times out or errors.

This module is pure logic — no LLM calls, no I/O. Caller provides:
  - the raw answer text (already had hallucinated tags stripped upstream)
  - the (tag, sentence) pairs that were verified
  - the verdicts list (aligned to pairs)
  - the policy name
  - the refusal_text string
"""

import logging
from typing import Any, Dict, List, Literal, Tuple

logger = logging.getLogger(__name__)

PolicyName = Literal["redact", "refuse", "warn"]
UNSUPPORTED_REDACTION_BN = "[তথ্য যাচাইযোগ্য নয়]"
_REFUSE_ESCALATION_FRACTION = 0.5


def apply_policy(
    answer: str,
    pairs: List[Tuple[str, str]],
    verdicts: List[str],
    policy: PolicyName,
    refusal_text: str,
) -> Tuple[str, List[Dict[str, Any]]]:
    """Apply the verifier policy. Returns (final_answer, events_to_emit).

    Events are dicts with `type` and `channel` set, ready to be yielded by the
    orchestrator. Event types used here: `unsupported_claim`, `verification_result`.

    The answer is mutated WITHOUT changing [S#] tags or the Sources block — the
    Sources block is appended AFTER policy runs, so this function only deals
    with the prose part of the answer.
    """
    if not pairs:
        return answer, [
            {"type": "verification_result", "channel": "debug",
             "verdicts": [], "policy": policy, "action": "noop_no_pairs"},
        ]

    events: List[Dict[str, Any]] = []

    not_entailed_count = sum(1 for v in verdicts if v == "not_entailed")
    partial_count = sum(1 for v in verdicts if v == "partial")
    entailed_count = sum(1 for v in verdicts if v == "entailed")

    summary_event = {
        "type": "verification_result",
        "channel": "debug",
        "policy": policy,
        "total": len(verdicts),
        "entailed": entailed_count,
        "partial": partial_count,
        "not_entailed": not_entailed_count,
    }

    # ----- WARN policy: keep answer, emit warnings only -----
    if policy == "warn":
        for (tag, sentence), v in zip(pairs, verdicts):
            if v != "entailed":
                events.append({
                    "type": "unsupported_claim",
                    "channel": "debug",
                    "tag": tag,
                    "sentence": sentence,
                    "verdict": v,
                    "action": "kept_warn_policy",
                })
        summary_event["action"] = "warn_only"
        events.append(summary_event)
        return answer, events

    # ----- REFUSE policy: any failure → full refusal -----
    if policy == "refuse":
        if not_entailed_count > 0:
            for (tag, sentence), v in zip(pairs, verdicts):
                if v == "not_entailed":
                    events.append({
                        "type": "unsupported_claim",
                        "channel": "debug",
                        "tag": tag,
                        "sentence": sentence,
                        "verdict": v,
                        "action": "triggered_refuse",
                    })
            summary_event["action"] = "full_refusal"
            events.append(summary_event)
            return refusal_text, events
        summary_event["action"] = "passed"
        events.append(summary_event)
        return answer, events

    # ----- REDACT policy (default) -----
    # Escalate to full refusal if too many sentences failed.
    if not_entailed_count and (not_entailed_count / max(len(verdicts), 1)) > _REFUSE_ESCALATION_FRACTION:
        for (tag, sentence), v in zip(pairs, verdicts):
            if v == "not_entailed":
                events.append({
                    "type": "unsupported_claim",
                    "channel": "debug",
                    "tag": tag,
                    "sentence": sentence,
                    "verdict": v,
                    "action": "triggered_escalation_to_refusal",
                })
        summary_event["action"] = "escalated_refusal"
        events.append(summary_event)
        return refusal_text, events

    # Sentence-level redaction: replace each not_entailed sentence with the
    # redaction marker. Use substring replace — sentences come straight from
    # extract_citations() which split on `।`/`.`/`!`/`?`, so they should be
    # unique substrings of the answer. If a sentence has been previously
    # redacted (duplicate not_entailed verdicts on the same sentence), the
    # second replace becomes a no-op.
    final = answer
    redacted_sentences = set()
    for (tag, sentence), v in zip(pairs, verdicts):
        if v == "not_entailed" and sentence not in redacted_sentences:
            if sentence and sentence in final:
                final = final.replace(sentence, UNSUPPORTED_REDACTION_BN)
                redacted_sentences.add(sentence)
            events.append({
                "type": "unsupported_claim",
                "channel": "debug",
                "tag": tag,
                "sentence": sentence,
                "verdict": v,
                "action": "redacted",
            })
        elif v == "partial":
            events.append({
                "type": "unsupported_claim",
                "channel": "debug",
                "tag": tag,
                "sentence": sentence,
                "verdict": v,
                "action": "kept_partial",
            })

    summary_event["action"] = (
        "redacted" if redacted_sentences else
        ("partial_only" if partial_count else "all_entailed")
    )
    events.append(summary_event)
    return final, events

File: mastra/src/verifier/policy.ts
/**
 * policy.ts — port of cogops/verifier/policy.py
 *
 * Turns NLI verdicts into a final answer + event list. Pure logic, no I/O.
 */

import { evt, type StreamEvent } from "../types.js";

export type PolicyName = "redact" | "refuse" | "warn";
export type Verdict = "entailed" | "partial" | "not_entailed";

export const UNSUPPORTED_REDACTION_BN = "[তথ্য যাচাইযোগ্য নয়]";
const REFUSE_ESCALATION_FRACTION = 0.5;

export function applyPolicy(
  answer: string,
  pairs: Array<[string, string]>,
  verdicts: Verdict[],
  policy: PolicyName,
  refusalText: string,
): [string, StreamEvent[]] {
  if (pairs.length === 0) {
    return [
      answer,
      [evt("verification_result", "debug", { verdicts: [], policy, action: "noop_no_pairs" })],
    ];
  }

  const events: StreamEvent[] = [];
  const notEntailed = verdicts.filter((v) => v === "not_entailed").length;
  const partial = verdicts.filter((v) => v === "partial").length;
  const entailed = verdicts.filter((v) => v === "entailed").length;

  const summary: Record<string, unknown> = {
    policy,
    total: verdicts.length,
    entailed,
    partial,
    not_entailed: notEntailed,
  };

  if (policy === "warn") {
    pairs.forEach(([tag, sentence], i) => {
      if (verdicts[i] !== "entailed") {
        events.push(
          evt("unsupported_claim", "debug", {
            tag,
            sentence,
            verdict: verdicts[i],
            action: "kept_warn_policy",
          }),
        );
      }
    });
    summary.action = "warn_only";
    events.push(evt("verification_result", "debug", summary));
    return [answer, events];
  }

  if (policy === "refuse") {
    if (notEntailed > 0) {
      pairs.forEach(([tag, sentence], i) => {
        if (verdicts[i] === "not_entailed") {
          events.push(
            evt("unsupported_claim", "debug", {
              tag,
              sentence,
              verdict: verdicts[i],
              action: "triggered_refuse",
            }),
          );
        }
      });
      summary.action = "full_refusal";
      events.push(evt("verification_result", "debug", summary));
      return [refusalText, events];
    }
    summary.action = "passed";
    events.push(evt("verification_result", "debug", summary));
    return [answer, events];
  }

  // --- REDACT (default) ---
  if (notEntailed && notEntailed / Math.max(verdicts.length, 1) > REFUSE_ESCALATION_FRACTION) {
    pairs.forEach(([tag, sentence], i) => {
      if (verdicts[i] === "not_entailed") {
        events.push(
          evt("unsupported_claim", "debug", {
            tag,
            sentence,
            verdict: verdicts[i],
            action: "triggered_escalation_to_refusal",
          }),
        );
      }
    });
    summary.action = "escalated_refusal";
    events.push(evt("verification_result", "debug", summary));
    return [refusalText, events];
  }

  let final = answer;
  const redactedSentences = new Set<string>();
  pairs.forEach(([tag, sentence], i) => {
    const v = verdicts[i];
    if (v === "not_entailed" && !redactedSentences.has(sentence)) {
      if (sentence && final.includes(sentence)) {
        final = final.split(sentence).join(UNSUPPORTED_REDACTION_BN);
        redactedSentences.add(sentence);
      }
      events.push(
        evt("unsupported_claim", "debug", { tag, sentence, verdict: v, action: "redacted" }),
      );
    } else if (v === "partial") {
      events.push(
        evt("unsupported_claim", "debug", { tag, sentence, verdict: v, action: "kept_partial" }),
      );
    }
  });

  summary.action =
    redactedSentences.size > 0 ? "redacted" : partial > 0 ? "partial_only" : "all_entailed";
  events.push(evt("verification_result", "debug", summary));
  return [final, events];
}
